File: frontend/oficina-web/src/app/models/ordem-servico.model.ts
import { BaseDTO } from './base.model';

export type PrioridadeOrdemServico = 'BAIXA' | 'NORMAL' | 'ALTA' | 'URGENTE';
export type StatusFluxoOrdemServico = 'ORCAMENTO' | 'EXECUCAO' | 'PAGAMENTO' | 'FINALIZADO';

export interface OrdemServicoResumo extends BaseDTO {
  numeroOs?: string;
  idCliente?: number;
  nomeCliente?: string;
  idVeiculo?: number;
  placaVeiculo?: string;
  descricaoVeiculo?: string;
  dataAbertura?: string;
  dataFinalizacao?: string;
  prioridade?: PrioridadeOrdemServico;
  valorTotal?: number;
  statusAtual?: string;
}

export interface AlterarStatusOrdemServico {
  novoStatus: StatusFluxoOrdemServico;
  observacao?: string;
}

export interface TotalRecursivoOrdemServico {
  idOrdemServico?: number;
  total?: number;
}

File: src/main/java/br/com/avcar/oficina/business/ordemservico/model/OrdemServicoModel.java
package br.com.avcar.oficina.business.ordemservico.model;

import br.com.avcar.oficina.business.ordemservico.enums.PrioridadeOrdemServico;
import br.com.avcar.oficina.business.pessoa.model.ClienteModel;
import br.com.avcar.oficina.business.veiculo.model.VeiculoModel;
import br.com.avcar.oficina.core.model.BaseModel;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import lombok.Getter;
import lombok.Setter;

/**
 * Registro principal do atendimento realizado pela oficina.
 *
 * Regra de negócio: toda Ordem de Serviço pertence a um único Cliente e a um
 * único Veículo, possui histórico de status e deve seguir o fluxo Orçamento,
 * Execução, Pagamento e Finalizado.
 */
@Getter
@Setter
@Entity
@Table(name = "ordem_servico")
public class OrdemServicoModel extends BaseModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_ordem_servico", nullable = false, updatable = false)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_cliente", nullable = false)
    private ClienteModel cliente;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_veiculo", nullable = false)
    private VeiculoModel veiculo;

    @Column(name = "numero_os", nullable = false, length = 30, unique = true)
    private String numeroOs;

    @Column(name = "data_abertura", nullable = false)
    private LocalDateTime dataAbertura = LocalDateTime.now();

    @Column(name = "data_aprovacao")
    private LocalDateTime dataAprovacao;

    @Column(name = "data_finalizacao")
    private LocalDateTime dataFinalizacao;

    @Enumerated(EnumType.STRING)
    @Column(name = "prioridade", nullable = false, length = 20)
    private PrioridadeOrdemServico prioridade = PrioridadeOrdemServico.NORMAL;

    @Column(name = "valor_total", nullable = false, precision = 12, scale = 2)
    private BigDecimal valorTotal = BigDecimal.ZERO;

    @Column(name = "observacao", columnDefinition = "TEXT")
    private String observacao;
}
